java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -68,300 +72,333 @@
+            result.add(new ExecutionState(executionState).setRelation(leftValue, operator, rightValue).push(SymbolicValue.BOOLEAN_TRUE));
+          }
+        }
       }
     }
 
     @Override
-    void evaluateConditionalAnd(BinaryExpressionTree tree) {
-      PackedStates leftResult = evaluateCondition(currentState, tree.leftOperand());
-      currentResult.falseStates.addAll(leftResult.falseStates);
-      for (ExecutionState state : leftResult.trueStates) {
-        PackedStates rightResult = evaluateCondition(state, tree.rightOperand());
-        currentResult.add(rightResult);
+    public final void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+      super.visitArrayAccessExpression(tree);
+      for (ExecutionState executionState : currentStates) {
+        executionState.pop();
+        executionState.pop();
+        executionState.push(createSymbolicInstanceValue());
       }
     }
 
     @Override
-    void evaluateConditionalOr(BinaryExpressionTree tree) {
-      PackedStates leftResult = evaluateCondition(currentState, tree.leftOperand());
-      currentResult.trueStates.addAll(leftResult.trueStates);
-      for (ExecutionState state : leftResult.falseStates) {
-        PackedStates rightResult = evaluateCondition(state, tree.rightOperand());
-        currentResult.add(rightResult);
+    public final void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      super.visitAssignmentExpression(tree);
+      for (ExecutionState executionState : currentStates) {
+        SymbolicValue expression = executionState.pop();
+        executionState.pop();
+        Symbol.VariableSymbol symbol = extractVariableSymbol(tree.variable());
+        if (symbol != null) {
+          executionState.assignValue(symbol, expression);
+        }
+        executionState.push(expression);
       }
     }
 
     @Override
-    void evaluateRelationalOperator(BinaryExpressionTree tree, SymbolicRelation operator) {
-      SymbolicValue leftValue = retrieveSymbolicValue(tree.leftOperand());
-      SymbolicValue rightValue = retrieveSymbolicValue(tree.rightOperand());
-      if (leftValue != null && rightValue != null) {
-        switch (currentState.evaluateRelation(leftValue, operator, rightValue)) {
-          case FALSE:
-            currentResult.falseStates.add(currentState);
+    public final void visitBinaryExpression(BinaryExpressionTree tree) {
+      Tree.Kind kind = ((JavaTree) tree).getKind();
+      SymbolicRelation relation = OPERATOR_TO_RELATION.get(kind);
+      if (relation != null) {
+        evaluateRelationalOperator(tree, relation);
+      } else {
+        switch (kind) {
+          case CONDITIONAL_AND:
+            evaluateConditionalAnd(tree);
             break;
-          case TRUE:
-            currentResult.trueStates.add(currentState);
+          case CONDITIONAL_OR:
+            evaluateConditionalOr(tree);
             break;
           default:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)